### PR TITLE
SAML IdP attribute mapping types and config handler

### PR DESCRIPTION
**Pull request by flyinghermit**
Supports attribute mapping configuration in `saml_idp_service_provider` spec.

```
 attribute_mapping:
  - name: string
    name_format: string
    value: string # predicate expression
```

value of `name_format` can be one of the following:
- unspecified or `urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified`. Used as a default value.
- uri or `urn:oasis:names:tc:SAML:2.0:attrname-format:uri`.
- basic or `urn:oasis:names:tc:SAML:2.0:attrname-format:basic`.

Reference spec file:
```
sp.yml
kind: saml_idp_service_provider
metadata:
  name: sp
  revision: 54eec4e0-1483-4b3e-807f-8421abd331de
spec:
  attribute_mapping:
  - name: displayname
    value: user.spec.traits.firstname
  - name: email
     name_format: urn
     value: user.spec.traits.email
  - name: firstname
    name_format: urn:oasis:names:tc:SAML:2.0:attrname-format:basic
    value: user.spec.traits.firstname
  entity_descriptor: |-
    <EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2023-12-10T16:26:51.083Z" entityID="https://example.com/saml/metadata">
      <SPSSODescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2023-12-10T16:26:51.083029Z" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol" AuthnRequestsSigned="false" WantAssertionsSigned="true">
        <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</NameIDFormat>
        <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://example.com/saml/acs" index="1"></AssertionConsumerService>
        <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://example.com/saml/acs" index="2"></AssertionConsumerService>
      </SPSSODescriptor>
    </EntityDescriptor>
  entity_id: https://example.com/saml/metadata
version: v1
```

RFD: https://github.com/gravitational/teleport.e/pull/2801
see https://github.com/gravitational/teleport/issues/34725, https://github.com/gravitational/teleport.e/issues/2692

changelog: attribute mapping in `saml_idp_service_provider` spec.

### CODE REVIEW

File: api/types/saml_idp_service_provider.go [hunk 1]
```diff
@@ -21,15 +21,25 @@ import (
 	"fmt"
 
 	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
 
 	"github.com/gravitational/teleport/api/utils"
 )
 
+const (
+	unspecifiedNameFormat = "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified"
```
Comment (flyinghermit) [inline api/types/saml_idp_service_provider.go:30]: We have these constants already defined in https://github.com/gravitational/teleport.e/blob/master/lib/idp/saml/assertion.go#L22. I am working on followup PRs which will likely move most (if not all) of the constants defined in that file to this one. So for now, I decided to just redefine the needed constants here. Let me know if you think that is better done in this PR.

**Comment by flyinghermit** [comment]
Friendly day 1 ping

File: api/types/saml_idp_service_provider.go (hunk 1 above)
Comment (r0mant) [reply api/types/saml_idp_service_provider.go:30]: As long as we don't end up in a permanent state where these constants are duplicated, I've no objections.

File: api/types/saml_idp_service_provider.go [hunk 2]
```diff
@@ -21,15 +21,25 @@ import (
...
+	uriNameFormat         = "urn:oasis:names:tc:SAML:2.0:attrname-format:uri"
+	basicNameFormat       = "urn:oasis:names:tc:SAML:2.0:attrname-format:basic"
+)
+
 var (
 	// ErrMissingEntityDescriptorAndEntityID is returned when both entity descriptor and entity ID is empty.
-	ErrEmptyEntityDescriptorAndEntityID = trace.BadParameter("either entity_descriptor or entity_id must be provided")
+	ErrEmptyEntityDescriptorAndEntityID = &trace.BadParameterError{Message: "either entity_descriptor or entity_id must be provided"}
```
Comment (r0mant) [inline api/types/saml_idp_service_provider.go:37]: Why did you change this, and errors below, from using `trace.BadParameter()` function?

File: api/types/saml_idp_service_provider.go [hunk 3]
```diff
@@ -161,6 +185,26 @@ func (s *SAMLIdPServiceProviderV1) CheckAndSetDefaults() error {
 		s.Spec.EntityID = ed.EntityID
 	}
 
+	if len(s.GetAttributeMapping()) > 0 {
+		attrNames := make([]string, 0)
```
Comment (r0mant) [inline api/types/saml_idp_service_provider.go:189]: Nit: Would probably be a little bit more efficient to use a map here instead of a slice.

File: api/types/saml_idp_service_provider.go [hunk 4]
```diff
@@ -161,6 +185,26 @@ func (s *SAMLIdPServiceProviderV1) CheckAndSetDefaults() error {
 		s.Spec.EntityID = ed.EntityID
 	}
 
+	if len(s.GetAttributeMapping()) > 0 {
```
Comment (r0mant) [inline api/types/saml_idp_service_provider.go:188]: This check is really redundant here. Just go straight to the "for" loop, and it will be a no-op if the mapping is empty.

File: api/types/saml_idp_service_provider.go [hunk 5]
```diff
@@ -161,6 +185,26 @@ func (s *SAMLIdPServiceProviderV1) CheckAndSetDefaults() error {
...
+			attrNames = append(attrNames, v.Name)
+
+			// validate input name format. If it is using one of the
+			// supported values "unspecified", "basic" or "uri", update
+			// the value with respective uri format.
+			uriNameFormat, err := validateAndGetURINameFormat(v.NameFormat)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			v.NameFormat = uriNameFormat
```
Comment (r0mant) [inline api/types/saml_idp_service_provider.go:204]: Can we define `CheckAndSetDefaults()` method on `SAMLAttributeMapping` that does this and call it here instead?

**Review by r0mant — COMMENTED** [review]
A few stylistic nitpicks.

File: api/types/saml_idp_service_provider.go (hunk 5 above)
Comment (flyinghermit) [reply api/types/saml_idp_service_provider.go]: https://github.com/gravitational/teleport/pull/35584/commits/50d6feb921a700f1e4c9ab54b77b6e3d035c067f

File: api/types/saml_idp_service_provider.go (hunk 2 above)
Comment (flyinghermit) [reply api/types/saml_idp_service_provider.go:36]: I received a suggestion in my backport that using `trace.BadParameter` directly will reference the location where it is defined instead of the line where the actual error occurs. But using `trace.BadParameterError` struct resolves that and prints a proper stack trace.

File: api/types/saml_idp_service_provider.go (hunk 1 above)
Comment (flyinghermit) [reply api/types/saml_idp_service_provider.go:29]: Thanks. Noted and will make sure we don't end up that way!

**Comment by flyinghermit** [comment]
Friendly ping @xacrimon @fspmarshall @mdwn

**Review by r0mant — APPROVED** [review]

**Review by mdwn — APPROVED** [review]
